File: app/src/main/java/com/kinwatt/powermeter/model/PowerProvider.java
package com.kinwatt.powermeter.model;

import android.content.Context;
import android.location.Location;

import com.kinwatt.powermeter.common.Function;
import com.kinwatt.powermeter.common.LocationUtils;
import com.kinwatt.powermeter.common.MathUtils;
import com.kinwatt.powermeter.sensor.LocationListener;
import com.kinwatt.powermeter.sensor.LocationProvider;

import java.util.ArrayList;
import java.util.List;

public class PowerProvider implements LocationListener {

    private static final int INTERPOLATION_STEP = 1000; // 1s

    private long baseTime;

    private LocationProvider locationProvider;
    private Location location1, location2;
    private Buffer<Location> buffer = new Buffer<>(10);

    private PowerAlgorithm powerAlgorithm;
    protected List<PowerListener> listeners = new ArrayList<>();

    public PowerProvider(Context context, LocationProvider locationProvider) {
        this.locationProvider = locationProvider;
        this.locationProvider.addListener(this);

        reset();
    }

    public PowerAlgorithm getPowerAlgorithm() {
        return powerAlgorithm;
    }
    public void setPowerAlgorithm(PowerAlgorithm powerAlgorithm) {
        this.powerAlgorithm = powerAlgorithm;
    }

    public void addListener(PowerListener listener) {
        listeners.add(listener);
    }
    public void removeListener(PowerListener listener) {
        listeners.remove(listener);
    }

    public void reset() {
        location1 = null;
        location2 = null;
        baseTime = 0;
        buffer.clear();
    }

    @Override
    public void onLocationChanged(Location location) {
        if (location2 == null) {
            baseTime = location.getTime();
        }
        location = new Location(location);
        location.setTime(location.getTime() - baseTime);

        if (location.getAltitude() == 0 && location1 != null && location2 != null) {
            softenAltitude(location);
        }

        if (location2 != null) {
            interpolatePositions(location);
        }
        else {
            buffer.add(location);
        }

        location1 = location2;
        location2 = location;
    }

    protected void onPowerCalculated(long time, float power) {
        for (PowerListener listener : listeners) {
            listener.onPowerMeasured(time, power);
        }
    }

    private void softenAltitude(Location position) {
        Function<Long, Double> altitude = MathUtils.interpolate(
                location1.getTime(), location1.getAltitude(),
                location2.getTime(), location2.getAltitude());

        position.setAltitude(altitude.apply(position.getTime()));

        if (position.getSpeed() == 0) {
            Function<Long, Float> speed = MathUtils.interpolate(
                    location1.getTime(), location1.getSpeed(),
                    location2.getTime(), location2.getSpeed());
            position.setSpeed(speed.apply(position.getTime()));
        }
    }

    private void interpolatePositions(Location position) {
        Function<Long, Location> interpolation = LocationUtils.interpolate(location2, position);

        long start = buffer.last().getTime() + INTERPOLATION_STEP;
        long end = position.getTime() + (INTERPOLATION_STEP - position.getTime() % INTERPOLATION_STEP);

        for (long i = start; i <= end; i += INTERPOLATION_STEP) {
            Location interpolatedPosition = interpolation.apply(i);
            Location lastPosition = buffer.last();

            float power = getPower(lastPosition, interpolatedPosition);

            buffer.add(interpolatedPosition);

            onPowerCalculated(interpolatedPosition.getTime() + baseTime, power);
        }
    }

    private float getPower(Location l1, Location l2) {
        //TODO: Calculate degrees from 5s ago;
        /*
        if (buffer.size() >= 5) {
            Location target = buffer.peek(4);

            double hDiff = l2.getAltitude() - target.getAltitude();
            double grade = hDiff / target.distanceTo(l2);

            return powerAlgorithm.calculatePower(l1, l2, grade);
        }
        else {
            return powerAlgorithm.calculatePower(l1, l2);
        }
        */
        return powerAlgorithm.calculatePower(l1, l2);
    }

    /*
    private static long getEndTime(long base, long target) {
        while (base < target) {
            base += INTERPOLATION_STEP;
        }
        return base;
    }
    */
}

File: app/src/main/java/com/kinwatt/powermeter/model/PowerProvider.kt
package com.kinwatt.powermeter.model

import android.location.Location

import com.kinwatt.powermeter.common.LocationUtils
import com.kinwatt.powermeter.sensor.LocationListener
import com.kinwatt.powermeter.sensor.LocationProvider
import java.util.*

class PowerProvider(var powerAlgorithm: PowerAlgorithm, private val locationProvider: LocationProvider) : LocationListener {

    private var baseTime: Long = 0
    private var previousLocation: Location? = null
    private val buffer = ArrayDeque<Location>(10)

    protected var listeners: MutableList<PowerListener> = ArrayList()

    init {
        this.locationProvider.addListener(this)

        reset()
    }

    fun addListener(listener: PowerListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: PowerListener) {
        listeners.remove(listener)
    }

    fun reset() {
        previousLocation = null
        baseTime = 0
        buffer.clear()
    }

    override fun onLocationChanged(location: Location) {
        if (previousLocation == null) baseTime = location.time

        val location = Location(location)
        location.time = location.time - baseTime

        if (previousLocation != null)
            interpolatePositions(location)
        else
            buffer.add(location)

        previousLocation = location
    }

    protected fun onPowerCalculated(time: Long, power: Float) {
        for (listener in listeners) {
            listener.onPowerMeasured(time, power)
        }
    }

    private fun interpolatePositions(position: Location) {
        val interpolation = LocationUtils.interpolate(previousLocation!!, position)

        val start = buffer.peekLast().time + INTERPOLATION_STEP
        val end = position.time + (INTERPOLATION_STEP - position.time % INTERPOLATION_STEP)

        for (i in start until end step INTERPOLATION_STEP) {
            val interpolatedPosition = interpolation(i)
            val lastPosition = buffer.peekLast()

            val power = getPower(lastPosition, interpolatedPosition)

            buffer.add(interpolatedPosition)

            onPowerCalculated(interpolatedPosition.time + baseTime, power)
        }
    }

    private fun getPower(l1: Location, l2: Location): Float {
        //TODO: Calculate degrees from 5s ago;
        /*
        if (buffer.size >= 5) {
            val target = buffer.elementAt(4)

            val hDiff = l2.altitude - target.altitude
            val grade = hDiff / target.distanceTo(l2)

            return powerAlgorithm.calculatePower(l1, l2, grade)
        } else {
            return powerAlgorithm.calculatePower(l1, l2)
        }
        */
        return powerAlgorithm.calculatePower(l1, l2)
    }

    companion object {
        protected val Logger = java.util.logging.Logger.getLogger("PowerProvider")!!

        private const val INTERPOLATION_STEP = 1000L // 1s

        private fun getEndTime(base: Long, target: Long): Long {
            var res = base
            while (res < target) {
                res += INTERPOLATION_STEP
            }
            return res
        }
    }
}
